Production Infrastructure

jacderida opened the pull request:
Hi Stephen/Calum,

This contains the setup for the production infrastructure for Jenkins. It runs on a VPC with a public and private subnet, with the Jenkins master being on the public subnet and the slaves being on the private one. There is also a 'Bastion' host on the public subnet to allow access to the internal network.

The infrastructure is entirely defined using Terraform. I will go back and replace the ad-hoc scripts for the development infrastructure to also use Terraform.

The process for getting it up and running is semi-automated. First the infrastructure is created, then we need to SSH to the Bastion and run the provisioning from in there. We could automate this, but perhaps it might not be used enough to justify doing so. The instructions are in the README.

Cheers,

Chris

jacderida committed:
feat/env: initial production infrastructure

The first attempt to describe the production infrastructure. We have a
VPC with a public subnet and a private subnet. The Jenkins master will
be in the public subnet and the slaves will be in the private subnet.
Eventually we'll be able to remove the definition of the slave machine
from this setup because we'll use the EC2Plugin for Jenkins that will
spin up the slaves dynamically.

The rules for the security groups setup the communication between the
public and private subnets. The rules need to be declared using
`aws_security_group_rule` resources, as opposed to using the
`aws_security_group` resource, because the groups reference each other,
and this creates a cycle in Terraform's acyclic graph.

jacderida committed:
feat/vm: machine for running ansible

This introduces a machine where we can run Ansible provisions from,
given that the other machines don't have direct SSH access. It seems
that communication between machines in different security groups on the
same subnet also need the rules to establish connectivity between the 2,
as with the public/private subnet communication. I don't understand
this, but they are definitely necessary for things to work correctly.

jacderida committed:
feat/vm: provision machines with ansible user

A user data script creates an ansible user for provisioning the
machines. It sets it up so it has passwordless sudo access. The Ansible
Bastion host can then be used as a provisioner and will have the ability
to connect to machines being provisioned as the ansible user.

jacderida committed:
feat/vm: ansible provisioner for prod infra

Adds some roles for setting up a machine to act as an Ansible
provisioner inside the VPC. This obviously needs to run from a machine
that already has Ansible on it. That could be a developer laptop or some
internal machine we eventually connect to Jenkins.

Right now this adds a private key to the Ansible Bastion host. This
should be removed and we should setup some kind of SSH tunnelling to
avoid having private keys on the Bastion host. For now we need this
setup for developing the prod infrastructure.

jacderida committed:
fix/role: install boto for ec2 provisions

The boto library is required by the ec2.py script. Newer versions of
Ubuntu don't appear to have a Python installation either, so this also
fixes that. It's reasonble for that to be in this role because it's
something required by Ansible.

jacderida committed:
fix/script: update based on distribution

On Ubuntu machines we also install Python, since newer versions don't
have an install by default.

There is also a sleep in the Makefile before provisioning the Bastion
host, as a general yum update runs before creating the ansible user.
This can be changed later to use a loop that tries to establish an SSH
connection.

jacderida committed:
feat/env: target to provision prod machines

This target is intended to be run from the Bastion host with the ansible
user, using the default key location.

jacderida committed:
fix/env: install java role

This installs the Java role on a fresh machine. I really need to get
Ansible Galaxy working again to handle this. There is also a fix here to
use the correct user for the provision.

jacderida committed:
fix/env: use private dns name for prod slaves

The prod slaves will not have public DNS names, so we use the internal
field instead. This also temporarily disables the Windows host as that's
not available yet.

Really we need two different configurations for the prod and dev
environments, or some other kind of way for the details to vary, as the
private hostnames may not work in the dev setup. I've added a task to
look at this later.

jacderida committed:
fix/vm: use ip rather than hostname

The internal hostname doesn't work. Also temporarily removed the Windows
host rather than commenting it out (which strangely didn't seem to
work).

jacderida committed:
feat/vm: open port 8080 for jenkins

jacderida committed:
fix/vm: introduce delay before yum update

For some reason this yum update seems to fail completetly randomly on
CentOS machines when it runs as a user data script. I think it needs
some more time to initialise or something.

jacderida committed:
feat/vm: rules for windows security group

This opens connectivity between the Bastion host and the Windows
machine, establishes internet connectivity for the Windows machine and
also opens ports for RDP and WinRM.

jacderida committed:
fix/role: install pypsrp in virtualenv

It turns out pypsrp needs to use a specific version of the requests
library, and ignoring the currently installed library doesn't work
correctly. This sets up a virtualenv and installs pypsrp in that. The
user performing the provision can then activate that virtualenv before
running it.

jacderida committed:
feat/vm: provision windows slave

This target is relying on the user setting the Jenkins URL and the
password for the windows machine. It would be possible to automate
getting these, but we can do it at a later point.

jacderida committed:
fix/vm: better variable name

It's misleading to refer to this as a URL and it could potentially
confuse someone running these scripts later. Change it to hostname,
since it's the hostname or the IP address you need to specify, not a
full URL.

jacderida committed:
fix/env: use correct environment and key

The Ansible user will have its own private key that will allow it to
connect to any target machines.

jacderida committed:
fix/vm: install python libraries in virtualenv

The pypsrp library wouldn't work correctly with a system-wide
installation because it requires a specific version of the requests
library. Ansible also needs to be installed in the virtualenv because
the ec2.py script references Ansible Python modules.

It's up to the user doing this provisioning to activate the virtualenv
before running the provision. We might automate this at some point in
the future, but don't want to invest time in this now.

jacderida committed:
feat/env: debugging for the prod environment

This introduces a debugging configuration for the production
environment. If you set DEBUG_JENKINS_ENV to true before calling the
target to create the environment, a Windows Bastion host will be
included in the setup. This will allow you to RDP to the Windows machine
on the private subnet. It is controlled with a count variable that is
set to 0 in the default configuration and 1 in the debugging
configuration.

jacderida committed:
fix/role: shell script to install pip packages

The 'virtualenv' parameter of the pip module didn't seem to work
correctly, so I'm just installing these 'manually' with a shell script
that activates the virtualenv before running the installation.

Also install the nc package for performing pre-flight checks before the
Ansible run.

jacderida committed:
fix/vm: try 2 minute pause

Both these machines, Linux and Windows, seem to be subject to this
problem where a download can fail in a user data script. I'm going to
try increasing the timeout to a couple of minutes to see if that might
help.

jacderida committed:
fix/role: use development version of ansible

Version 2.7.9 doesn't have the fix for the Chocolatey non-zero exit
code.

jacderida committed:
feat/env: open port 8080 for Windows slave group

This also re-adds the Windows slave definition to the Jenkins
configuration. It was taken out while the Windows part wasn't working
yet.

jacderida committed:
fix/env: open port 50000 for slave connectivity

jacderida committed:
docs/env: setup production jenkins on AWS

Commit:
chore/Makefile:add terraform init command

calumcraig reviewed (dismissed):
Was able to work through the steps outlined https://github.com/jacderida/safe-build-infrastructure/tree/production_infrastructure#aws-production-provision to successfully provision this environment:
![image](https://user-images.githubusercontent.com/44327924/55016090-ffd89580-4fe5-11e9-9ecd-724176833820.png)

jacderida committed:
docs/env: correct and clarify instructions

These were based on review feedback from Calum going through the
instructions.

jacderida committed:
Merge branch 'production_infrastructure' of github.com:jacderida/safe-build-infrastructure into production_infrastructure

calumcraig reviewed (dismissed):
These look good to me - as discussed during the process of testing this PR these were points I encountered.

jacderida committed:
docs/env: explicitly wait for prov to finish

This makes it a bit clearer that this whole process has to complete
before you do the reset.

S-Coyle reviewed (approved):
Ran through the instructions on macOS 10.14.3 and successfully provisioned the production environment. 👍

jacderida committed:
Merge remote-tracking branch 'upstream/master' into production_infrastructure
